### PR TITLE
Create gold image

### PR DESCRIPTION
**Pull request by 14rcole**
Create gold image template and Dockerfile.  The Gold image is a Jenkins image off of which other CentOS-PaaS-SIG jenkins master images will be based

### CODE REVIEW

**Comment by 14rcole** [comment]
@p3ck PTAL

File: config/s2i/gold/jenkins-persistent-buildconfig-template.yaml [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+- description: Git source URI for Jenkins S2I
+  name: REPO_URL
+  required: true
+  value: https://github.com/CentOS-PaaS-SIG/ci-pipeline.git
+- description: Git branch/tag reference
+  name: REPO_REF
+  value: master
+- description: Path within Git project to build; empty for root project directory.
+  name: CONTEXT_DIR
+  value: config/s2i/jenkins/master
```
Comment (p3ck) [inline config/s2i/gold/jenkins-persistent-buildconfig-template.yaml]: Is this the correct path?  Wouldn't this be the path to config/s2i/gold/jenkins/master ?

File: config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    strategy:
+      dockerStrategy:
+      type: Docker
+    triggers:
+    - type: ConfigChange
+parameters:
+- description: Git repository with Dockerfile and slave entrypoint.
+  displayName: Repository URL
+  name: REPO_URL
+  value: https://github.com/CentOS-PaaS-SIG/ci-pipeline.git
```
Comment (johnbieren) [inline config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml]: Should this point to this repo?

File: config/s2i/gold/jenkins-persistent-buildconfig-template.yaml (hunk 1 above)
Comment (14rcole) [reply config/s2i/gold/jenkins-persistent-buildconfig-template.yaml]: Yes, thank you

File: config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+- description: The sub-directory inside the repository.
+  displayName: Context Directory
+  name: CONTEXTDIR
+  value: config/s2i/jenkins/slave
```
Comment (johnbieren) [inline config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml]: To echo what Bill said, I believe this should have gold in the path to mirror where this file is being proposed to be added

File: config/s2i/gold/jenkins-persistent-buildconfig-template.yaml (hunk 1 above)
Comment (johnbieren) [reply config/s2i/gold/jenkins-persistent-buildconfig-template.yaml]: +1

File: config/s2i/gold/jenkins-persistent-buildconfig-template.yaml [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+  value: jenkins-jnlp
+- name: ENABLE_OAUTH
+  displayName: Enable OAuth in Jenkins
+  description: Whether to enable OAuth OpenShift integration. If false, the static
+    account 'admin' will be initialized with the password 'password'.
+  value: 'true'
+- name: MEMORY_LIMIT
+  displayName: Memory Limit
+  description: Maximum amount of memory the container can use.
+  value: 512Mi
```
Comment (johnbieren) [inline config/s2i/gold/jenkins-persistent-buildconfig-template.yaml]: @p3ck thoughts? I know I always up this. If this is 'gold,' shouldn't it be a value we don't all have to up? To the best of my knowledge, you up it too.

File: config/s2i/gold/jenkins-persistent-buildconfig-template.yaml [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+- name: VOLUME_CAPACITY
+  displayName: Volume Capacity
+  description: Volume space available for data, e.g. 512Mi, 2Gi.
+  value: 1Gi
```
Comment (johnbieren) [inline config/s2i/gold/jenkins-persistent-buildconfig-template.yaml]: Same as above @p3ck

File: config/s2i/gold/master/configuration/init.groovy [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+// Set global and job read permissions
+def strategy = Jenkins.instance.getAuthorizationStrategy()
+strategy.add(hudson.model.Hudson.READ,'anonymous')
+strategy.add(hudson.model.Item.READ,'anonymous')
+Jenkins.instance.setAuthorizationStrategy(strategy)
+// Set Markup Formatter to Safe HTML so PR hyperlinks work
+Jenkins.instance.setMarkupFormatter(new RawHtmlMarkupFormatter(false))
+Jenkins.instance.save()
+
+logger.info("Setup fedora-fedmsg Messaging Provider")
```
Comment (johnbieren) [inline config/s2i/gold/master/configuration/init.groovy]: I don't see jms messaging in the plugins.txt. Unsure how this works without the plugin installed. Not sure if there is magic behind it and it will like add the plugin for you or not (not sarcastic, actually don't know what will happen here).

File: config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
+apiVersion: v1
```
Comment (johnbieren) [inline config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml]: Nothing in continuous-infra-create-openshift-project seems to call this file

File: config/s2i/gold/jenkins-persistent-buildconfig-template.yaml (hunk 4 above)
Comment (p3ck) [reply config/s2i/gold/jenkins-persistent-buildconfig-template.yaml]: Yes - We have found that 512MB is way too low.  We should move this to 2GB

File: config/s2i/gold/jenkins-persistent-buildconfig-template.yaml (hunk 5 above)
Comment (p3ck) [reply config/s2i/gold/jenkins-persistent-buildconfig-template.yaml]: Agreed - move this to 10Gi

File: config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml (hunk 2 above)
Comment (p3ck) [reply config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml]: +1

File: config/s2i/gold/master/configuration/init.groovy (hunk 6 above)
Comment (14rcole) [reply config/s2i/gold/master/configuration/init.groovy]: I don't know either.  That init.groovy is either used in or a part of a larger init.groovy for each project, but not all projects have the jms-messaging plugin.  I don't see anything wrong with adding it.

File: config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml (hunk 3 above)
Comment (johnbieren) [reply config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml]: I guess what I am saying is: Nothing in this file is pointing to this current repo so I am not sure why it is being added here without being modified to be hosted in this repo. If the idea is not to have a gold slave image living here, should this file be here?

File: config/s2i/gold/master/configuration/init.groovy (hunk 6 above)
Comment (johnbieren) [reply config/s2i/gold/master/configuration/init.groovy]: This init.groovy code adds a fedmsg message provider to the master. I am pretty sure it depends on jms-messaging being present. I would be surprised if there was a master with this code, but without jms-messaging as a plugin. Without jms-messaging, I expect this to do nothing or fail, just not sure which

File: config/s2i/gold/continuous-infra-create-openshift-project [hunk 8]
```diff
@@ -0,0 +1,19 @@
+PROJECT=continuous-infra
```
Comment (arilivigni) [inline config/s2i/gold/continuous-infra-create-openshift-project]: This file is not necessary we already have contra-env-setup to take the gold images and load them

File: config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml (hunk 2 above)
Comment (arilivigni) [reply config/s2i/gold/jenkins-continuous-infra-slave-buildconfig-template.yaml]: +1

**Comment by 14rcole** [comment]
@arilivigni in that case, all I really need to add is a Dockerfile to build config/s2i/jenkins/master, right?  Unless the plugins.txt needs to contain the combination of all packages used by various jenkins masters instead of their intersection as I had been doing

**Comment by 14rcole** [comment]
@arilivigni the init.groovy that already exists in contra-env-sample-project is more extenstive than the other init.groovy files.  Do I proceed with the more extensive init.groovy file or use the minimal one?

Also, this plugins.txt contains plugins unique to contra-env-sample-project.  I assume I need to keep those, but should I include other unique plugins (e.g. plugins that only linchpin uses) as well?

**Comment by p3ck** [comment]
this should be a superset of the common plugins and settings for all our current projects.  Only leave things out if they shouldn't be upstream or they conflict.

**Comment by 14rcole** [comment]
@p3ck Does that include jobs?

**Comment by p3ck** [comment]
I would say no to jobs.

**Comment by 14rcole** [comment]
@arilivigni Can you take a look?

**Comment by arilivigni** [comment]
@14rcole please have other review first I am in an all day meeting

File: config/s2i/jenkins/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,7 @@
+FROM openshift/jenkins-2-centos7
```
Comment (arilivigni) [inline config/s2i/jenkins/Dockerfile]: Is this file necessary? Doesn't plugins get installed regardless?

Comment (14rcole) [reply config/s2i/jenkins/Dockerfile]: They do, but if a project wants to use this image and has other plugins to install, plugins.txt will be overridden.

File: config/s2i/jenkins/master/configuration/scriptApproval.xml [hunk 10]
```diff
@@ -5,11 +5,25 @@
...
+        
+        <string>method java.lang.Class isInstance java.lang.Object</string>
+        <string>method org.jenkinsci.plugins.workflow.support.actions.EnvironmentAction getEnvironment</string>
+        <string>new java.lang.Exception java.lang.String java.lang.Throwable</string>
+        <string>new java.lang.Exception java.lang.String</string>
+        <string>staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods contains java.lang.Object[] java.lang.Object</string>
+        <string>staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods getAt java.util.List groovy.lang.Range</string>
+        <string>staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods getAt java.util.List java.util.Collection</string>
+        <string>staticMethod org.jenkinsci.plugins.pipeline.modeldefinition.Utils markStageSkippedForConditional java.lang.String</string>
+
```
Comment (johnbieren) [inline config/s2i/jenkins/master/configuration/scriptApproval.xml]: Looks like a copy paste error here with the extra line and tab

**Comment by joejstuart** [comment]
lgtm

File: config/s2i/jenkins/Dockerfile (hunk 9 above)
Comment (arilivigni) [reply config/s2i/jenkins/Dockerfile]: @14rcole same question

Comment (14rcole) [reply config/s2i/jenkins/Dockerfile]: They do, but if a project wants to use this image and has additional plugins to install this allows plugins.txt to be overridden

Comment (14rcole) [reply config/s2i/jenkins/Dockerfile]: I'll have to go back and look again, but at the time I created this file, I think I understood the requirement to be a minimal base that functions as a starting point for linchpin, ci-pipeline, etc. master images.  Is this more of a hybrid of everything the jenkins masters could possibly need?

Comment (arilivigni) [reply config/s2i/jenkins/Dockerfile]: Jenkins Core needs to be updated and with the set of plugins that we use is fine but it still needs updating regardless.
Does that make sense?

Comment (14rcole) [reply config/s2i/jenkins/Dockerfile]: Yeah

File: config/s2i/jenkins/master/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,9 @@
+FROM openshift/jenkins-2-centos7:latest
```
Comment (arilivigni) [inline config/s2i/jenkins/master/Dockerfile]: we don't need this anymore since we are pulling the OpenShift image now correct?

Comment (14rcole) [reply config/s2i/jenkins/master/Dockerfile]: You're right, fixed now

**Comment by arilivigni** [comment]
@joejstuart can you review as well?
